### PR TITLE
feat: create generic callback support for PyTorch

### PR DESCRIPTION
**Pull request by yoavz**
First phase of https://determinedai.atlassian.net/browse/DET-3033 to add support for a generic callback interface in PyTorchTrial. Second phase will involve fixing the `on_validation_step_end()` case in the distributed / parallel context.

Test Plan:
Added unit tests.

### CODE REVIEW

File: harness/determined/pytorch/_callback.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+    .. warning::
+        If you are defining a stateful callback (e.g. it mutates a self
+        attribute over it's lifetime), you must also override state_dict() and
+        load_state_dict() to ensure this state can be serialized and deserialized
+        over checkpoints.
+
+    .. warning::
+        If distributed or parallel training is enabled, every shard will
+        execute a copy of this callback (except for on_validation_step_end).
```
[the remark this answers is not in the document]
Comment (yoavz) [reply harness/determined/pytorch/_callback.py:19]: Are you referring to `on_validation_step_end`? The executing `on_validation_step_end` on a single GPU (as opposed to all shards) is intended to be temporary, while we aren't doing the validation metrics reduction yet. The follow-up PR will do the reduction in validation steps and remove this warning.

File: harness/determined/pytorch/_callback.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        ``trial.context.distributed.get_rank()``.
+    """
+
+    def on_train_step_start(self, step_id: int) -> None:
+        """
+        Run before every training step begins.
+        """
+        pass
+
+    def on_train_step_end(self, step_id: int, metrics: Dict[str, Any]) -> None:
```
[the remark this answers is not in the document]
Comment (yoavz) [reply harness/determined/pytorch/_callback.py:31]: Oh, didn't realize we didn't always average training metrics. Will add this comment.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 3]
```diff
@@ -738,6 +777,33 @@ def build_validation_data_loader(self) -> DataLoader:
...
+            optimizer (torch.optim.Optimizer): instance of the optimizer to be
+                used for training
+
+        Returns:
+            :py:class:`det.pytorch.LRScheduler`:
+                Wrapper around a :obj:`torch.optim.lr_scheduler._LRScheduler`.
+        """
+        pass
+
+    def build_callbacks(self) -> Dict[str, _callback.PyTorchCallback]:
```
[the remark this answers is not in the document]
Comment (yoavz) [reply harness/determined/pytorch/_pytorch_trial.py:797]: See `_save()` implementation -- we need a way to link the restored callback state to the correct callback, if there are multiple. If we used a list, we would have to depend on the order of the callback list which would get messy.

**Comment by yoavz** [comment]
@aaron276h I thought a bit about the `context` available to the callback and realized that you could actually use the following pattern: 

```
class MyTrial(...):
    def __init__(self, context: context):
        self.optimizer = ...
        self.model = ....
        self.some_callback = Callback(self.model, self.optimizer)

    def build_model(self ,...):
        return self.model

    def build_optimizer(self, ...):
        return self.optimizer

    def build_callback(self, ...)
        return {"some_callback": self.some_callback})
```

In a sense, this is closer to our existing API anyways -- any shared state the trial class implementations need to use can be initialized in the constructor, e.g. how one would share state between data loaders and model implementation. 

I figured this would be a better place to start than exposing the context object, which is a different approach than our existing plugin functions take. What do you think?

File: harness/determined/pytorch/_callback.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    .. warning::
+        If distributed or parallel training is enabled, every shard will
+        execute a copy of this callback (except for on_validation_step_end).
+        To configure a callback implementation to execute on a subset of shards,
+        please condition your implementation on
+        ``trial.context.distributed.get_rank()``.
+    """
+
+    def on_train_step_start(self, step_id: int) -> None:
```
[the remark this answers is not in the document]
Comment (yoavz) [reply harness/determined/pytorch/_callback.py:25]: Yes, that option is left up to the user. see my top-level comment.

File: harness/determined/pytorch/_callback.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply harness/determined/pytorch/_callback.py:19]: Ok yeah than this makes sense for now.

File: harness/determined/pytorch/_pytorch_trial.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (aaron276h) [reply harness/determined/pytorch/_pytorch_trial.py:797]: If we keep it as a list why would would it get messy, since the return of `build_callbacks` is immutable across pause and restart? I just find the return type `Dict[str, _callback.PyTorchCallback]` a bit strange.

**Comment by yoavz** [comment]
> If we keep it as a list why would would it get messy, since the return of build_callbacks is immutable across pause and restart?

continue training is the main use-case i'm thinking about. what if you want to continue training and stop using the lr scheduler, for instance?

**Comment by yoavz** [comment]
> @yoavz The problem with this is that for multi-GPU training, self.optimizer here would be pointing to the wrong optimizer and we don't have a clean way for them access the updated optimizer, besides creating a context object.

got it, then I will add the optimizer, model, and scheduler to the context.

File: harness/determined/pytorch/_callback.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+    the lifetime of a PyTorchTrial.
+
+    .. warning::
+        If you are defining a stateful callback (e.g. it mutates a self
+        attribute over it's lifetime), you must also override state_dict() and
+        load_state_dict() to ensure this state can be serialized and deserialized
+        over checkpoints.
+
+    .. warning::
+        If distributed or parallel training is enabled, every shard will
```
Comment (aaron276h) [inline harness/determined/pytorch/_callback.py:18]: In our other docs, I don't believe we use the term `shard`, for consistency might be better change it to GPU.

File: harness/determined/pytorch/_pytorch_context.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+class PyTorchTrialContext(det.TrialContext):
+    """
+    Base context class that contains runtime information for any Determined
+    workflow that uses the ``pytorch`` API.
+    """
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+        # The following three attributes are initialized during the lifetime of a
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_context.py:20]: This comment shouldn't be here

File: harness/determined/pytorch/_pytorch_context.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+import torch
+import torch.nn as nn
+
+import determined as det
+from determined import pytorch
+from determined_common import check
+
+
+class PyTorchTrialContext(det.TrialContext):
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_context.py:11]: Do we need a `PytorchClientContext`?

File: harness/determined/pytorch/_callback.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
+from typing import Any, Dict
+
+import torch
+
+
+class PyTorchCallback:
+    """
+    Abstract base class used to define a callback that should execute during
+    the lifetime of a PyTorchTrial.
```
Comment (aaron276h) [inline harness/determined/pytorch/_callback.py:9]: I think we should add a bit of documentation here that users should access callbacks via `context` object.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 9]
```diff
@@ -137,19 +143,23 @@ def _init_model(self) -> None:
...
-        self.lr_helper = _LRHelper(self.trial.create_lr_scheduler(self.optimizer))
+        self.context.model = self.model
 
-        # If a load path is provided load weights and restore the data location.
-        self._load()
-
-        self._configure_amp()
+    def _init_optimizer(self) -> None:
+        # TODO: Check that optimizer is not an amp optimizer.
+        self.optimizer = self.trial.optimizer(self.model)
```
Comment (aaron276h) [inline harness/determined/pytorch/_pytorch_trial.py:150]: non-blocking: now that we have `self.context` we probably shouldn't have `self.optimizer` and `self.model` anymore.

File: harness/determined/pytorch/_pytorch_context.py (hunk 6 above)
Comment (yoavz) [reply harness/determined/pytorch/_pytorch_context.py:20]: Updated.

File: harness/determined/pytorch/_pytorch_context.py (hunk 7 above)
Comment (yoavz) [reply harness/determined/pytorch/_pytorch_context.py:11]: We do not have a native API for pytorch yet, so nope.

File: harness/determined/pytorch/_callback.py (hunk 8 above)
Comment (yoavz) [reply harness/determined/pytorch/_callback.py:9]: The intention of the API is for users to access these through the getters, which are documented.

File: harness/determined/pytorch/_pytorch_trial.py (hunk 9 above)
Comment (yoavz) [reply harness/determined/pytorch/_pytorch_trial.py]: done
